### PR TITLE
8299807: newStringNoRepl should avoid copying arrays for ASCII compatible charsets

### PR DESCRIPTION
**Pull request by Glavo**
This is the javadoc of `JavaLangAccess::newStringNoRepl`:

```java
    /**
     * Constructs a new {@code String} by decoding the specified subarray of
     * bytes using the specified {@linkplain java.nio.charset.Charset charset}.
     *
     * The caller of this method shall relinquish and transfer the ownership of
     * the byte array to the callee since the later will not make a copy.
     *
     * @param bytes the byte array source
     * @param cs the Charset
     * @return the newly created string
     * @throws CharacterCodingException for malformed or unmappable bytes
     */
```

It is recorded in the document that it should be able to directly construct strings with parameter byte array to reduce array allocation.

However, at present, `newStringNoRepl` always copies arrays for UTF-8 or other ASCII compatible charsets.

This PR fixes this problem.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299807](https://bugs.openjdk.org/browse/JDK-8299807): newStringNoRepl should avoid copying arrays for ASCII compatible charsets

### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12119/head:pull/12119` \
`$ git checkout pull/12119`

Update a local copy of the PR: \
`$ git checkout pull/12119` \
`$ git pull https://git.openjdk.org/jdk pull/12119/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12119`

View PR using the GUI difftool: \
`$ git pr show -t 12119`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12119.diff">https://git.openjdk.org/jdk/pull/12119.diff</a>

</details>

### CODE REVIEW

**Comment by Glavo** [comment]
I ran the tier1 and tier2 tests, and there were no new errors.

**Comment by Glavo** [comment]
The only use case affected is `Files.readString`. I tested the performance of `readString` based on the memory file system.

baseline:
```
Benchmark                   (length)   Mode  Cnt        Score       Error  Units
NoRepl.testReadAscii               0  thrpt    5  5049760.744 ±  3563.324  ops/s
NoRepl.testReadAscii            1024  thrpt    5  3523083.785 ± 23747.078  ops/s
NoRepl.testReadAscii            8192  thrpt    5  2415952.140 ± 85884.289  ops/s
NoRepl.testReadAscii         1048576  thrpt    5    32425.563 ±   284.121  ops/s
NoRepl.testReadAscii        33554432  thrpt    5      872.492 ±     4.311  ops/s
NoRepl.testReadAscii       268435456  thrpt    5       58.736 ±     0.224  ops/s
NoRepl.testReadAsciiAsGBK          0  thrpt    5  4229547.832 ± 10997.381  ops/s
NoRepl.testReadAsciiAsGBK       1024  thrpt    5  3409472.580 ±   819.566  ops/s
NoRepl.testReadAsciiAsGBK       8192  thrpt    5  2179865.886 ± 16606.862  ops/s
NoRepl.testReadAsciiAsGBK    1048576  thrpt    5    32962.429 ±   249.385  ops/s
NoRepl.testReadAsciiAsGBK   33554432  thrpt    5      871.810 ±     1.812  ops/s
NoRepl.testReadAsciiAsGBK  268435456  thrpt    5       59.131 ±     0.172  ops/s
NoRepl.testReadGBK                 0  thrpt    5  4657464.074 ± 51849.667  ops/s
NoRepl.testReadGBK              1024  thrpt    5  1199083.242 ±  8653.846  ops/s
NoRepl.testReadGBK              8192  thrpt    5    75823.949 ±    46.493  ops/s
NoRepl.testReadGBK           1048576  thrpt    5      675.214 ±     1.729  ops/s
NoRepl.testReadGBK          33554432  thrpt    5       20.972 ±     2.280  ops/s
NoRepl.testReadGBK         268435456  thrpt    5        2.585 ±     0.394  ops/s
NoRepl.testReadUTF8                0  thrpt    5  5222403.558 ± 44728.740  ops/s
NoRepl.testReadUTF8             1024  thrpt    5  1417472.161 ± 23776.534  ops/s
NoRepl.testReadUTF8             8192  thrpt    5   185714.265 ±  2096.328  ops/s
NoRepl.testReadUTF8          1048576  thrpt    5     1318.763 ±    16.051  ops/s
NoRepl.testReadUTF8         33554432  thrpt    5       30.663 ±     0.114  ops/s
NoRepl.testReadUTF8        268435456  thrpt    5        3.782 ±     0.041  ops/s
```

This PR:
```
Benchmark                   (length)   Mode  Cnt        Score       Error  Units
NoRepl.testReadAscii               0  thrpt    5  5084610.141 ± 37449.826  ops/s
NoRepl.testReadAscii            1024  thrpt    5  3425713.961 ± 24060.542  ops/s
NoRepl.testReadAscii            8192  thrpt    5  2765684.248 ± 20586.103  ops/s
NoRepl.testReadAscii         1048576  thrpt    5    48074.603 ±   371.213  ops/s
NoRepl.testReadAscii        33554432  thrpt    5     1167.878 ±    14.427  ops/s
NoRepl.testReadAscii       268435456  thrpt    5       71.028 ±     0.439  ops/s
NoRepl.testReadAsciiAsGBK          0  thrpt    5  4783174.805 ±  9789.109  ops/s
NoRepl.testReadAsciiAsGBK       1024  thrpt    5  3518265.840 ± 18467.577  ops/s
NoRepl.testReadAsciiAsGBK       8192  thrpt    5  2775108.822 ± 19282.776  ops/s
NoRepl.testReadAsciiAsGBK    1048576  thrpt    5    46956.963 ±   147.593  ops/s
NoRepl.testReadAsciiAsGBK   33554432  thrpt    5     1165.036 ±    10.032  ops/s
NoRepl.testReadAsciiAsGBK  268435456  thrpt    5       70.878 ±     0.191  ops/s
NoRepl.testReadGBK                 0  thrpt    5  4910043.054 ± 27295.344  ops/s
NoRepl.testReadGBK              1024  thrpt    5  1177675.970 ± 15573.239  ops/s
NoRepl.testReadGBK              8192  thrpt    5    75417.479 ±   233.957  ops/s
NoRepl.testReadGBK           1048576  thrpt    5      674.620 ±     5.856  ops/s
NoRepl.testReadGBK          33554432  thrpt    5       19.899 ±     1.504  ops/s
NoRepl.testReadGBK         268435456  thrpt    5        2.705 ±     0.002  ops/s
NoRepl.testReadUTF8                0  thrpt    5  4851516.950 ±  9237.743  ops/s
NoRepl.testReadUTF8             1024  thrpt    5  1332016.420 ±  9570.465  ops/s
NoRepl.testReadUTF8             8192  thrpt    5   184177.766 ±  4662.562  ops/s
NoRepl.testReadUTF8          1048576  thrpt    5     1326.439 ±     3.420  ops/s
NoRepl.testReadUTF8         33554432  thrpt    5       30.782 ±     0.116  ops/s
NoRepl.testReadUTF8        268435456  thrpt    5        3.790 ±     0.011  ops/s
```

When reading an ASCII file as UTF-8 or GBK encoding, we can see that the throughput of `readString` has improved significantly. When reading ASCII files with a size of 1MiB, the throughput increased by 40%~50%, but for larger or smaller files, the throughput improvement will be smaller. 

For files containing non-ASCII characters, the throughput of `readString` is between 94% and 104% of the baseline.

This is the source code of the benchmark: https://gist.github.com/Glavo/f3d2060d0bd13cd0ce2add70e6060ea0

**Comment by Glavo** [comment]
Can someone help me open an Issue on Java Bug System?

**Comment by Glavo** [comment]
The throughput of reading ASCII files as UTF-8:

image


The throughput of reading ASCII files as GBK:

image

**Comment by RogerRiggs** [comment]
It seems odd that the benchmark seems slower for smaller files; can you suggest why that might be?
I'd expect the size distribution for Files.readString to be biased toward the smaller files.
Can you repeat the benchmark using the default file system.  OS file caching should eliminate the disk speed effects.

**Comment by Glavo** [comment]
@RogerRiggs 

I rerun benchmark based on the default file system, and the test file size is between 0 and 32KiB.

The throughput of reading ASCII files as UTF-8:
image

The throughput of reading ASCII files as GBK:
image

The performance has been slightly improved, and there is no performance degradation.

For UTF-8 and GBK files with non-ASCII characters, the throughput fluctuates by no more than 4%.

Test code and original results: https://gist.github.com/Glavo/f3d2060d0bd13cd0ce2add70e6060ea0?permalink_comment_id=4451350#gistcomment-4451350

**Comment by Glavo** [comment]
> It seems odd that the benchmark seems slower for smaller files; can you suggest why that might be?

The most likely reason is the cost of the newly added if judgment in newStringUTF8NoRepl.

I don't think this is an important issue, because when it comes to actual I/O operations, its impact is negligible.

The main purpose of this PR is to eliminate unnecessary temporary memory allocation, thus reducing GC pressure. The change in throughput is only a by-product.

**Comment by Glavo** [comment]
> /issue JDK-8299807

Thank you!

File: src/java.base/share/classes/java/lang/String.java [hunk 1]
```diff
@@ -698,7 +698,7 @@ private String(Charset charset, byte[] bytes, int offset, int length) {
     /*
      * Throws iae, instead of replacing, if malformed or unmappable.
      */
-    static String newStringUTF8NoRepl(byte[] bytes, int offset, int length) {
+    static String newStringUTF8NoRepl(byte[] bytes, int offset, int length, boolean noCopy) {
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/lang/String.java:701]: The parameter will be easier to understand if `true` means that the string created must not share the byte array supplied.
For example,

\@param noShare {@code true} if the resulting string MUST NOT share the byte array, {@code false} if the byte array can be exclusively used to construct the string and is not modified or used for any other purpose.

This flips the boolean argument in the uses below.

Comment (Glavo) [reply src/java.base/share/classes/java/lang/String.java]: > The parameter will be easier to understand if `true` means that the string created must not share the byte array supplied. For example,
> 
> @param noShare {@code true} if the resulting string MUST NOT share the byte array, {@code false} if the byte array can be exclusively used to construct the string and is not modified or used for any other purpose.
> 
> This flips the boolean argument in the uses below.

I modified it according to your suggestion.

**Comment by Glavo** [comment]
Benchmark results for large files (128KiB~256MiB) on the default file system:

image

image

For files of about 1MiB, the throughput can be improved by more than 55%.

Original results: https://gist.github.com/Glavo/f3d2060d0bd13cd0ce2add70e6060ea0?permalink_comment_id=4452798#gistcomment-4452798

**Comment by Glavo** [comment]
Don't turn it off

**Comment by Glavo** [comment]
Does anyone review this PR again?

**Comment by Glavo** [comment]
@RogerRiggs I have revised it according to your suggestion. Can you have a look again?

**Review by RogerRiggs — APPROVED** [review]
Looks good, thanks for the updates.

**Pull request closed without merging**